File: examples/sendEmail.mjs
import { TraceClient } from "../sdk/packages/client/dist/index.js";
import { sendEmail } from "./lib/mailer.mjs";

const trace = new TraceClient({ endpoint: "http://localhost:8787" });

const decision = await trace.propose({
  type: "send_email",
  actor: { kind: "agent", name: "mail-bot", provider: "openai" },
  target: "mailto:sarah@acme.com",
  params: { subject: "Pricing", body: "Hi!" }
});

console.log("Decision:", decision);

if (decision.status === "requires_approval") {
  await fetch(`http://localhost:8787/approve/${decision.actionId}`, { method: "POST" });
  console.log("Approved:", decision.actionId);
}

try {
  const result = await sendEmail({
    to: "sarah@acme.com",
    subject: "Pricing",
    body: "Hi!"
  });

  await trace.evidence(decision.actionId, [
    { name: "email_sent", pass: true, note: `msgId=${result.id}` }
  ]);
  console.log("Evidence submitted (success).");
} catch (err) {
  await trace.evidence(decision.actionId, [
    { name: "email_failed", pass: false, note: String(err?.message ?? err) }
  ]);
  console.log("Evidence submitted (failure).");
}

File: examples/python/send_email.py
from trace_client import TraceClient, Actor
import requests

trace = TraceClient(endpoint="http://localhost:8787")

# 1) Propose an action
decision = trace.propose(
    type="send_email",
    actor=Actor(kind="agent", name="mail-bot", provider="openai"),
    target="mailto:sarah@acme.com",
    params={"subject": "Pricing", "body": "Hi!"}
)
print("Decision:", decision["status"], decision["checks"])

# 2) (Dev) simulate approval if required
if decision["status"] == "requires_approval":
    requests.post(f"http://localhost:8787/approve/{decision['actionId']}")
    print("Approved:", decision["actionId"])

# 3) Perform your side-effect (fake mailer here)
def send_email(to: str, subject: str, body: str):
    # simulate a send and return metadata
    return {"id": "msg_demo_123", "to": to, "subject": subject}

result = send_email("sarah@acme.com", "Pricing", "Hi!")

# 4) Attach evidence (success)
trace.evidence(decision["actionId"], [
    {"name": "email_sent", "pass": True, "note": f"id={result['id']}"}
])
print("Evidence submitted.")
